test/unit/state_resolver_test.rb

```diff
@@ -0,0 +1,171 @@
+            when "no" then question :a
+            end
+          end
+        end
+
+        radio :y do
+          option :yes
+          option :no
+          next_node do |response|
+            case response
+            when "yes" then outcome :a
+            when "no" then outcome :b
+            end
+          end
+        end
+        outcome :a
+        outcome :b
+      end
+
+      @state_resolver = StateResolver.new(@flow)
+    end
+
+    context "#state_from_response_store" do
+      should "return the state for the first question when there are no responses" do
+        response_store = ResponseStore.new(responses: {})
+        state = @state_resolver.state_from_response_store(response_store)
+
+        assert_equal :x, state.current_node
+        assert_empty state.accepted_responses
+        assert_nil state.current_response
+      end
+
+      should "return the state for the next question when there are valid answers" do
+        response_store = ResponseStore.new(responses: { "x" => "yes" })
+        state = @state_resolver.state_from_response_store(response_store)
+
+        assert_equal :y, state.current_node
+        assert_equal ({ x: "yes" }), state.accepted_responses
+        assert_nil state.current_response
+      end
+
+      should "return an error state for the current question when an answer is invalid" do
+        response_store = ResponseStore.new(responses: { "x" => "yes", "y" => "invalid" })
+        state = @state_resolver.state_from_response_store(response_store)
+
+        assert_equal :y, state.current_node
+        assert_equal ({ x: "yes" }), state.accepted_responses
+        assert_equal "invalid", state.current_response
+        assert state.error.present?
+      end
+
+      should "return an outcome state when all questions are answered" do
+        response_store = ResponseStore.new(responses: { "x" => "yes", "y" => "no" })
+        state = @state_resolver.state_from_response_store(response_store)
+
+        assert_equal :b, state.current_node
+        assert_equal ({ x: "yes", y: "no" }), state.accepted_responses
+        assert_nil state.current_response
+      end
+
+      should "return an particular node if one is requested" do
+        response_store = ResponseStore.new(responses: { "x" => "yes", "y" => "no" })
+        state = @state_resolver.state_from_response_store(response_store, "y")
+
+        assert_equal :y, state.current_node
+        assert_equal ({ x: "yes" }), state.accepted_responses
+        assert_equal "no", state.current_response
+      end
+
+      should "return an error state for an earlier node regardless of a requested node" do
+        response_store = ResponseStore.new(responses: { "x" => "invalid", "y" => "no" })
+        state = @state_resolver.state_from_response_store(response_store, "y")
+
+        assert_equal :x, state.current_node
+        assert_equal ({}), state.accepted_responses
+        assert_equal "invalid", state.current_response
```

Comment:
Should we be checking if there's an error? `assert state.error.present?`

Reply:
Oh yes, thanks for spotting.